test/test_autograd.py

```diff
@@ -5761,6 +5773,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+            self.assertEqual(a, y.grad_fn._saved_other)
+            y.sum().backward()
+            self.assertEqual(2 * a, a.grad)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: 2 * x, lambda x: x / 2)
+            a = torch.randn(5, requires_grad=True)
+            y = a * a
+            self.assertEqual(a, y.grad_fn._saved_self)
+            self.assertEqual(a, y.grad_fn._saved_other)
+            y.sum().backward()
+            self.assertEqual(2 * a, a.grad)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: 2 * x, lambda x: x)
+            a = torch.randn(5, requires_grad=True)
+            y = a * a
+            self.assertEqual(2 * a, y.grad_fn._saved_self)
+            self.assertEqual(2 * a, y.grad_fn._saved_other)
+            y.sum().backward()
+            self.assertEqual(4 * a, a.grad)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+        # Exited hooks correctly
+        a = torch.randn(5, requires_grad=True)
+        y = a * a
+        self.assertEqual(a, y.grad_fn._saved_self)
+        self.assertEqual(a, y.grad_fn._saved_other)
+        y.sum().backward()
+        self.assertEqual(2 * a, a.grad)
+
+    def test_default_saved_variable_hooks_did_not_save_original(self):
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+            a = torch.randn(5, requires_grad=True)
+            y = torch.exp(a)
+            self.assertEqual(y, y.grad_fn._saved_result)
+            y.sum().backward()
+            self.assertEqual(a.grad, y)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+    def test_setting_default_saved_variable_hooks_twice_should_fail(self):
+        try:
+            with self.assertRaisesRegex(RuntimeError, "Setting default hooks but they have already been set. "):
+                torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+                torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+    def test_saving_variable_to_disk(self):
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            def pack(x):
+                name = os.path.join(tmp_dir, str(uuid.uuid4()))
+                torch.save(x, name)
+                return name
+
+            def unpack(name):
+                return torch.load(name)
+
+            try:
+                torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
+                a = torch.ones(5, requires_grad=True)
+                y = a * a
+                self.assertEqual(a, y.grad_fn._saved_self)
```

Comment:
Is the ability to access SavedVariables on the grad_fn a new feature? Is this a public feature? (judging by the underscore, seems like no).

The reason for the question is that I'm worried about what happens if someone tries to mutate _saved_self or change it. For example,
```
y.grad_fn._saved_self.zero_()
y.grad_fn._saved_self = torch.randn(3, 3)
```

Reply:
This was added by @soulitzer with #52451. If the user mutates `y.grad_fn._saved_self`, it has the same behavior as mutating directly `a`.

If `a` is a leaf, it will fail early with `RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.`
If not, it will fail when the variable is unpacked, thanks to the version counter logic. (`RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation:`)